presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java

```diff
@@ -177,18 +213,60 @@ private void enforceTimeLimits()
-            Duration queryMaxExecutionTime = getQueryMaxExecutionTime(query.getSession());
+            QueryLimit<Duration> queryMaxExecutionTime = getMinimum(
+                    createDurationLimit(getQueryMaxExecutionTime(query.getSession()), QUERY),
+                    query.getResourceGroupQueryLimits()
+                            .flatMap(ResourceGroupQueryLimits::getExecutionTimeLimit)
+                            .map(rgLimit -> createDurationLimit(rgLimit, RESOURCE_GROUP)).orElse(null));
             Optional<DateTime> executionStartTime = query.getExecutionStartTime();
             DateTime createTime = query.getCreateTime();
-            if (executionStartTime.isPresent() && executionStartTime.get().plus(queryMaxExecutionTime.toMillis()).isBeforeNow()) {
-                query.fail(new PrestoException(EXCEEDED_TIME_LIMIT, "Query exceeded the maximum execution time limit of " + queryMaxExecutionTime));
+            if (executionStartTime.isPresent() && executionStartTime.get().plus(queryMaxExecutionTime.getLimit().toMillis()).isBeforeNow()) {
+                query.fail(
+                        new PrestoException(EXCEEDED_TIME_LIMIT,
+                                format(
+                                        "Query exceeded the maximum execution time limit of %s defined at the %s level",
+                                        queryMaxExecutionTime.getLimit(),
+                                        queryMaxExecutionTime.getLimitSource().name())));
             }
             if (createTime.plus(queryMaxRunTime.toMillis()).isBeforeNow()) {
                 query.fail(new PrestoException(EXCEEDED_TIME_LIMIT, "Query exceeded maximum time limit of " + queryMaxRunTime));
             }
         }
     }
 
+    /**
+     *  When cluster reaches max tasks limit and also a single query
+     *  exceeds a threshold,  kill this query
+     */
+    private void enforceTaskLimits()
+    {
+        int totalRunningTaskCount = 0;
+        int highestRunningTaskCount = 0;
+        Optional<T> highestRunningTaskQuery = Optional.empty();
+        for (T query : queries.values()) {
+            if (query.isDone() || !(query instanceof QueryExecution)) {
+                continue;
+            }
+            int runningTaskCount = ((QueryExecution) query).getRunningTaskCount();
+            totalRunningTaskCount += runningTaskCount;
+            if (runningTaskCount > highestRunningTaskCount) {
+                highestRunningTaskCount = runningTaskCount;
+                highestRunningTaskQuery = Optional.of(query);
+            }
+        }
+
+        runningTaskCount.set(totalRunningTaskCount);
+
+        if (totalRunningTaskCount > maxTotalRunningTaskCountToKillQuery &&
+                highestRunningTaskCount > maxQueryRunningTaskCount &&
+                highestRunningTaskQuery.isPresent()) {
+            highestRunningTaskQuery.get().fail(new PrestoException(QUERY_HAS_TOO_MANY_STAGES, format(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: totalRunningTaskCount (defined at line 243), highestRunningTaskCount (defined at line 244). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 263. Consider formatting this data with &#39;%d&#39; instead.